pink/limits/floating_base_velocity_limit.py

```diff
@@ -0,0 +1,156 @@
+            max_linear_velocity: Linear velocity limits along the base frame
+                axes. A scalar applies the same bound to all axes.
+            max_angular_velocity: Angular velocity limits around the base frame
+                axes. A scalar applies the same bound to all axes.
+        """
+        self.model = model
+        self.base_frame = base_frame
+        self.linear_max = _as_velocity_vector(
+            max_linear_velocity, "max_linear_velocity"
+        )
+        self.angular_max = _as_velocity_vector(
+            max_angular_velocity, "max_angular_velocity"
+        )
+        if model.existJointName("root_joint"):
+            self.root_joint_id = model.getJointId("root_joint")
+            root_joint = model.joints[self.root_joint_id]
+            self.root_idx_v = root_joint.idx_v
+            self.root_nv = root_joint.nv
+        else:
+            self.root_joint_id = None
+            self.root_idx_v = None
+            self.root_nv = 0
+
+        if model.existFrame(base_frame):
+            self.frame_id = model.getFrameId(base_frame)
+            parent_joint = model.frames[self.frame_id].parentJoint
+        else:
+            self.frame_id = None
+            parent_joint = None
+
+        if self.root_joint_id is not None and self.frame_id is None:
+            raise ValueError(
+                f"Frame '{base_frame}' does not exist in the model."
+            )
+
+        if (
+            self.frame_id is not None
+            and self.root_joint_id is not None
+            and parent_joint != self.root_joint_id
+        ):
+            raise ValueError(
+                f"Frame '{base_frame}' is not attached to the root joint."
+            )
+
+    def compute_qp_inequalities(
+        self,
+        configuration: "Configuration",
+        dt: float,
+    ) -> Optional[Tuple[np.ndarray, np.ndarray]]:
+        """Linearize floating base velocity bounds."""
+        if (
+            self.frame_id is None
+            or self.root_joint_id is None
+            or not np.isfinite(self.linear_max).any()
+            and not np.isfinite(self.angular_max).any()
+        ):
+            return None
+
+        jacobian = pin.getFrameJacobian(
+            self.model,
+            configuration.data,
+            self.frame_id,
+            pin.ReferenceFrame.LOCAL,
+        )
+
+        rows = []
+        bounds = []
+
+        for axis, bound in enumerate(self.linear_max):
+            if not np.isfinite(bound):
+                continue
+            row = jacobian[axis, :].copy()
+            if self.root_idx_v is not None:
+                row[: self.root_idx_v] = 0.0
+                after_start = self.root_idx_v + self.root_nv
+                if after_start < row.shape[0]:
+                    row[after_start:] = 0.0
+            rows.extend([row, -row])
+            bounds.extend([dt * bound, dt * bound])
+
+        for axis, bound in enumerate(self.angular_max):
```

Comment:
We could combine these two loops into a single one.

```py
  for bound_vector, offset in [(self.linear_max, 0), (self.angular_max, 3)]:
      for axis, bound in enumerate(bound_vector):
          if not np.isfinite(bound):
              continue
          ...
```

Better yet, we could vectorize this entirely to avoid loops:

- Rather than `self.linear_max` and `self.angular_max`, we could store `self.twist_max = [linear_max, angular_max]`
- Instead of processing row by row, start from `J = jacobian`, set `J[:, :self.root_idx_v] = 0.0` and similarly `J[:, after_start:] = 0.0`
- G becomes a vstack of `[J, -J]`
- h becomes an hstack of `[self.twist_max * dt, -self.twist_max * dt]`

There would be no loops then and the code will run faster. What do you think?